### ISSUE TITLE
8BitDo Mod Kit for Original N64 Controller

### ISSUE THREAD

**Issue by Keffkaa** [Mr-Wiseguy/Zelda64Recomp#158]
**Describe the bug**
Controller inputs are not recognized. Works in Windows and on any N64 emulator. It's in Dinput mode. Only way to use it was to use JoytoKey to remap keyboard inputs to the controller but then I only get 8 way directions (as if I'm using WASD)

**To Reproduce**
1. Go to Controller settings.
2. Try to remap any button and nothing happens. 

**Expected behavior**
Controller should work, or at least be able to change bindings. It's as if it didn't exist. 

https://shop.8bitdo.com/products/8bitdo-mod-kit-for-original-n64-controller

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Version: 23H2
 - CPU: Ryzen 7800X3d
 - GPU: 4070ti

**Comment by bkacjios** [comment]
I use an 8bitdo modkit with steam. Just setup all the buttons through steam then rebind them through the controls menu and it works fine.

**Comment by SvenBudak** [comment]
> I use an 8bitdo modkit with steam. Just setup all the buttons through steam then rebind them through the controls menu and it works fine.

Rumble also work?

**Comment by Drihscol** [comment]
> > I use an 8bitdo modkit with steam. Just setup all the buttons through steam then rebind them through the controls menu and it works fine.
> 
> Rumble also work?

I'm using a modkit through steam, rumble still works

**Comment by SvenBudak** [comment]
Hmm, okay, nevertheless... I would recommend supporting this controller natively without additional programs or similar tools and implementing proper controller mapping in n64recomp. After all, there is no alternative, and anyone who wants to experience a true N64 experience will opt for this controller.

**Comment by bkacjios** [comment]
> > > I use an 8bitdo modkit with steam. Just setup all the buttons through steam then rebind them through the controls menu and it works fine.
> > 
> > 
> > Rumble also work?
> 
> I'm using a modkit through steam, rumble still works

@Drihscol Are you using Switch or D-input mode? I actually wasn't able to get rumble working, so I'm curious.

> Hmm, okay, nevertheless... I would recommend supporting this controller natively without additional programs or similar tools and implementing proper controller mapping in n64recomp. After all, there is no alternative, and anyone who wants to experience a true N64 experience will opt for this controller.

@SvenBudak yes, I agree. It would be nice if it just worked. Steam is just a work around for now at least.

**Comment by SvenBudak** [comment]
> yes, I agree. It would be nice if it just worked. Steam is just a work around for now at least.

Honestly, I don't understand what you mean by stream ^^ I have installed stream, but N64Recomp is a standalone .exe ... I don't start that via stream. Or are you talking about a different stream (not the game launcher from Valve)?

**Comment by bkacjios** [comment]
Yes I am talking about Steam, the game library platform made by Valve. You can add third party games to steam via `Games -> Add a Non-Steam Game to My Library` 

You can then launch the game through steam and allow you to use a custom controller profile.

![image](https://github.com/Zelda64Recomp/Zelda64Recomp/assets/3247233/371f93dc-8a46-46e9-b5e2-da54d8ee6b8a)

**Comment by Drihscol** [comment]
> > > > I use an 8bitdo modkit with steam. Just setup all the buttons through steam then rebind them through the controls menu and it works fine.
> > > 
> > > 
> > > Rumble also work?
> > 
> > 
> > I'm using a modkit through steam, rumble still works
> 
> @Drihscol Are you using Switch or D-input mode? I actually wasn't able to get rumble working, so I'm curious.
> 
> > Hmm, okay, nevertheless... I would recommend supporting this controller natively without additional programs or similar tools and implementing proper controller mapping in n64recomp. After all, there is no alternative, and anyone who wants to experience a true N64 experience will opt for this controller.
> 
> @SvenBudak yes, I agree. It would be nice if it just worked. Steam is just a work around for now at least.

I use Switch mode :)

**Comment by SvenBudak** [comment]
Hmm... i will use then my regular 5€ PC Controller for now ^^"

**Comment by bkacjios** [comment]
> I use Switch mode :)

Hmm.. I just tried switch mode and rumble works now, but c-right is also acting as d-pad right AND opening the setting menu. That not happening for you?

**Comment by Mr-Wiseguy** [comment]
As I can't provide support myself due to not having an 8BitDo modkit (and not wanting to use a modkit on one of my N64 controllers), I'd need somebody to contribute this support to the project. There may be a way to add support with the current version of SDL by using a custom gamepad mapping, but as stated I wouldn't be able to test that myself. [SDL_GameControllerDB](https://github.com/mdqinc/SDL_GameControllerDB) does seem to have some [entries](https://github.com/mdqinc/SDL_GameControllerDB/blob/9055df09a2bdd56339096639168c8bcdc6c0edc2/gamecontrollerdb.txt#L37) named "8BitDo N64", but I'm not sure if they're referring to this modkit.

**Comment by bkacjios** [comment]
Ah right, I forgot about custom mapping for SDL. I can give it a try tomorrow!

**Comment by bkacjios** [comment]
I can confirm the UID for my mod kit is `03000000c82d00006928000000000000` when in D-Input mode and `030000007e0500001920000000000000` when in switch mode, which both seem to be in the gamecontrollerdb.txt you suggested.


```
03000000c82d00006928000000000000,8BitDo N64,a:b0,b:b1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b12,leftshoulder:b6,lefttrigger:b8,leftx:a0,lefty:a1,rightshoulder:b7,righttrigger:b9,rightx:a3,righty:a4,start:b11,platform:Windows,
030000007e0500001920000000000000,NSO N64 Controller,+rightx:b8,+righty:b2,-rightx:b3,-righty:b7,a:b1,b:b0,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b12,leftshoulder:b4,lefttrigger:b6,leftx:a0,lefty:a1,misc1:b13,rightshoulder:b5,righttrigger:b10,start:b9,platform:Windows,
```

I put that gamecontrollerdb.txt next to Zelda64Recompiled.exe but the inputs are still not working. I even tried out the gamepad mapping tool and confirmed this entry is correct.

DInput mode
![gamepad-tool_FnCDgvZGAO](https://github.com/Zelda64Recomp/Zelda64Recomp/assets/3247233/d053b221-6521-443b-b8eb-133459770711)

Switch mode
![gamepad-tool_LguVWNJDyD](https://github.com/Zelda64Recomp/Zelda64Recomp/assets/3247233/f4616355-3048-4039-92c3-ca40b4700539)

**Comment by Mr-Wiseguy** [comment]
Making a gamecontrollerdb.txt file next to the executable won't automatically load it, you also need to set the `SDL_GAMECONTROLLERCONFIG` environment variable to point at the file. It looks like the gamepad mapping tool has a "Set Mapping As Environment Variable" button, I'm not sure exactly what that does but it sounds like it'd set `SDL_GAMECONTROLLERCONFIG` for you.

**Comment by bkacjios** [comment]
Huh, yeah, that actually worked. Controller works fine now in game.

**Comment by paparansen** [comment]
> Huh, yeah, that actually worked. Controller works fine now in game.

would you please post the steps on how to use the sdl gamepad mapper?
i tried some stuff but no luck. the tool detects the controller, but 2 of the
yellow buttons are not recognized, also cant get games to work with it.
